DIPs/DIP10xx.md

```diff
@@ -0,0 +1,117 @@
+
+## Contents
+* [Rationale](#rationale)
+* [Prior Work](#prior-work)
+* [Description](#description)
+* [Reference](#reference)
+* [Copyright & License](#copyright--license)
+* [Reviews](#reviews)
+
+## Rationale
+
+You may often need to change the content of a static array.
+This feature will save you some time, especially for `betterC` users (because
+dynamic arrays are not available), as you won't need to change the array's length anymore.
+
+## Prior Work
+
+- The `$` operator, which is well implemented.
+- [PR 3615](https://github.com/dlang/dmd/pull/3615)
+- [PR 590](https://github.com/dlang/dlang.org/pull/590)
+- [Issue 14070](https://issues.dlang.org/show_bug.cgi?id=14070)
+- [Issue 8008](https://issues.dlang.org/show_bug.cgi?id=8008)
+
+## Description
+
+This feature will be compatible with `betterC`.
+
+Actually, a static array must be declared like this:
+
+```d
+uint[2] arr = [ 2, 4, ];
+```
+
+If we want to make it more flexible at compile time, but we still want it static, we can't.
+For example:
+```d
+string str;
+foreach (arg; args)
+   str ~= arg ~ ", "
+mixin("int[] arr = [" ~ str ~ "];");    // The array is not static
+```
+
+The `$` into the brackets will be replaced, at compile time, by the array's length.
+
+```d
+//...
+string str;
+foreach (arg; args)
+   str ~= arg ~ ", ";
+mixin("int[$] arr = [" ~ str ~ "];");    // The array is static
+```
+
+This feature will ONLY be available for declarations.
+
+```d
+int[$] a1 = [1, 2, 3];              // OK, `$` = 3
+char[$] a2 = "cerise";              // OK
+auto a3 = cast(int[$]) [1, 2, 3];   // OK, `a3` is a int[3] (same behaviour as `cast(int[3])`
+
+int[$][] a4 = [[1, 2], [3, 4]];     // OK, same as `int[2][]`
+foreach (int[$] a; a4) {}           // OK, we can deduce `$` because it's length (2) is fixed in `a4`
+
+int[][] a5 = [[1, 2, 3], [4, 5, 6]];
+foreach (int[$] a; a5) {}           // Error: `a5` is a dynamic array
+
+const[$] a6 = [6, 7, 8];            // Error: basic type expected, not [
+auto[$] a7 = [6, 7, 8];             // Error: basic type expected, not [
+
+int[$] a8;                          // Error: length of static array `a8` cannot be deduced. Please replace `$` by a number.
+
+void foo(int[$] p) {}               // Error: length of static array `p` cannot be deduced. Please replace `$` by a number.
+
+int[$] bar(int[2] arr)              // OK, `$` = 4
+{
+    return arr ~ [3, 4];
+}
+
+int[$] baz(int[] arr)               // Error: Cannot deduce the length of the return value of function `baz`.
+{
+    return arr;
+}
+```
+
+### Grammar changes
+
+[declaration spec](https://dlang.org/spec/declaration.html)
+```diff
+ int[3] x;     // x is an array of 3 ints
+ int[3][5] x;  // x is an array of 5 arrays of 3 ints
+ int[3]*[5] x; // x is an array of 5 pointers to arrays of 3 ints
++int[$] x = [ 0, 1 ]; // x is a static array of 2 ints
++int[$][] x = [[0, 2], [1, 3]]; // x is a dynamic array of static arrays with a length of 2
+```
```

Comment:
This is not quite a formal grammar change, but I guess it's okay for this DIP.

Reply:
(I have no idea how to write a grammar change)